File: Game-1-modular/systems/skill_unlock_system.py
"""Skill unlock system for managing skill availability"""

from typing import List, Optional, Set, TYPE_CHECKING

from data.models.skill_unlocks import SkillUnlock
from data.databases import SkillUnlockDatabase

if TYPE_CHECKING:
    from entities.character import Character


class SkillUnlockSystem:
    """
    Manages skill unlock progression for a character.

    Tracks:
    - Which skills have been unlocked
    - Which skills are currently unlockable (conditions met, awaiting payment)
    - Provides trigger-based checking for new unlocks
    """

    def __init__(self):
        self.unlocked_skills: Set[str] = set()  # skill_ids that have been unlocked
        self.pending_unlocks: Set[str] = set()  # unlock_ids awaiting cost payment
        self.unlock_db = SkillUnlockDatabase.get_instance()

    def is_skill_unlocked(self, skill_id: str) -> bool:
        """Check if a skill has been unlocked."""
        return skill_id in self.unlocked_skills

    def is_unlock_pending(self, unlock_id: str) -> bool:
        """Check if an unlock is pending cost payment."""
        return unlock_id in self.pending_unlocks

    def check_for_unlocks(self, character: 'Character', trigger_type: Optional[str] = None,
                          trigger_value: any = None) -> List[SkillUnlock]:
        """
        Check for new skill unlocks based on character state.

        Args:
            character: Character instance with full state access
            trigger_type: Optional trigger filter (level_up, title_earned, etc.)
            trigger_value: Optional trigger value (level, title_id, etc.)

        Returns:
            List of SkillUnlock definitions that became available
        """
        newly_unlockable = []

        for unlock in self.unlock_db.get_all_unlocks():
            # Skip if already unlocked
            if unlock.skill_id in self.unlocked_skills:
                continue

            # Skip if already pending
            if unlock.unlock_id in self.pending_unlocks:
                continue

            # If trigger filter specified, check if this unlock matches
            if trigger_type and unlock.trigger.type != trigger_type:
                continue

            # If trigger value specified, check if it matches
            if trigger_value is not None and unlock.trigger.trigger_value != trigger_value:
                continue

            # Check if conditions are met
            if not unlock.check_conditions(character):
                continue

            # Conditions met! Check if there's a cost
            can_afford, _ = unlock.check_cost(character)

            if unlock.cost.gold == 0 and len(unlock.cost.materials) == 0 and unlock.cost.skill_points == 0:
                # No cost - unlock immediately
                success, message = self._unlock_skill(character, unlock)
                if success:
                    newly_unlockable.append(unlock)
            elif can_afford:
                # Has cost but can afford - mark as pending for player confirmation
                self.pending_unlocks.add(unlock.unlock_id)
                newly_unlockable.append(unlock)
            else:
                # Has cost but can't afford yet - conditions met but not unlockable
                pass

        return newly_unlockable

    def unlock_skill(self, character: 'Character', unlock_id: str) -> tuple[bool, str]:
        """
        Unlock a skill by unlock_id (e.g., from player UI interaction).

        Args:
            character: Character unlocking the skill
            unlock_id: ID of the unlock to perform

        Returns:
            (success, message)
        """
        unlock = self.unlock_db.get_unlock(unlock_id)
        if not unlock:
            return False, f"Unknown unlock: {unlock_id}"

        # Check if already unlocked
        if unlock.skill_id in self.unlocked_skills:
            return False, "Skill already unlocked"

        # Attempt unlock
        success, message = self._unlock_skill(character, unlock)
        if success:
            # Remove from pending if it was there
            self.pending_unlocks.discard(unlock_id)

        return success, message

    def _unlock_skill(self, character: 'Character', unlock: SkillUnlock) -> tuple[bool, str]:
        """
        Internal method to unlock a skill.

        Args:
            character: Character unlocking the skill
            unlock: SkillUnlock definition

        Returns:
            (success, message)
        """
        # Final check: can unlock?
        can_unlock, reason = unlock.can_unlock(character)
        if not can_unlock:
            return False, reason

        # Pay cost and unlock
        success, message = unlock.unlock(character)
        if success:
            self.unlocked_skills.add(unlock.skill_id)
            return True, message

        return False, message

    def get_pending_unlocks(self) -> List[str]:
        """Get list of unlock_ids that are pending."""
        return list(self.pending_unlocks)

    def check_level_up_unlocks(self, character: 'Character', new_level: int) -> List[SkillUnlock]:
        """
        Check for unlocks triggered by level up.

        Args:
            character: Character that leveled up
            new_level: New level reached

        Returns:
            List of newly unlockable skills
        """
        return self.check_for_unlocks(character, trigger_type='level_up', trigger_value=new_level)

    def check_title_earned_unlocks(self, character: 'Character', title_id: str) -> List[SkillUnlock]:
        """
        Check for unlocks triggered by earning a title.

        Args:
            character: Character that earned the title
            title_id: ID of the title earned

        Returns:
            List of newly unlockable skills
        """
        return self.check_for_unlocks(character, trigger_type='title_earned', trigger_value=title_id)

    def check_quest_complete_unlocks(self, character: 'Character', quest_id: str) -> List[SkillUnlock]:
        """
        Check for unlocks triggered by quest completion.

        Args:
            character: Character that completed the quest
            quest_id: ID of the completed quest

        Returns:
            List of newly unlockable skills
        """
        return self.check_for_unlocks(character, trigger_type='quest_complete', trigger_value=quest_id)

    def check_activity_threshold_unlocks(self, character: 'Character') -> List[SkillUnlock]:
        """
        Check for unlocks triggered by activity thresholds.

        This is called periodically (e.g., after crafting, combat, gathering)
        to check milestone-based unlocks.

        Args:
            character: Character to check

        Returns:
            List of newly unlockable skills
        """
        return self.check_for_unlocks(character, trigger_type='activity_threshold')

    def get_unlocked_count(self) -> int:
        """Get number of skills unlocked."""
        return len(self.unlocked_skills)

    def reset_debug(self):
        """Reset unlock state (for debug/testing)."""
        self.unlocked_skills.clear()
        self.pending_unlocks.clear()

File: Unity/Assets/Scripts/Game1.Systems/Progression/SkillUnlockSystem.cs
// ============================================================================
// Game1.Systems.Progression.SkillUnlockSystem
// Migrated from: systems/skill_unlock_system.py (206 lines)
// Migration phase: 4
// Date: 2026-02-13
//
// Manages skill unlock progression. Skills unlock via triggers:
//   level_up, title_earned, quest_complete, activity_threshold.
// Skills with no cost unlock immediately; those with a cost become pending
// until the player confirms the purchase.
// ============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace Game1.Systems.Progression
{
    // ========================================================================
    // Skill Unlock Definition
    // ========================================================================

    /// <summary>
    /// Trigger definition for when a skill can become unlockable.
    /// </summary>
    public class UnlockTrigger
    {
        /// <summary>Trigger type: "level_up", "title_earned", "quest_complete", "activity_threshold".</summary>
        public string Type { get; set; }

        /// <summary>Trigger value (e.g., level number, title_id, quest_id).</summary>
        public object TriggerValue { get; set; }
    }

    /// <summary>
    /// Cost to unlock a skill (gold, materials, skill points).
    /// </summary>
    public class UnlockCost
    {
        public int Gold { get; set; }
        public int SkillPoints { get; set; }
        public List<QuestItemRequirement> Materials { get; set; } = new();

        /// <summary>Check if there is any cost at all.</summary>
        public bool HasCost => Gold > 0 || SkillPoints > 0 || Materials.Count > 0;
    }

    /// <summary>
    /// Full skill unlock definition. Loaded from skill unlock JSON.
    /// Matches Python SkillUnlock dataclass.
    /// </summary>
    public class SkillUnlockDefinition
    {
        public string UnlockId { get; set; }
        public string SkillId { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public UnlockTrigger Trigger { get; set; }
        public UnlockCost Cost { get; set; } = new();

        /// <summary>
        /// Check if character meets the conditions for this unlock.
        /// Delegates to character state interface.
        /// </summary>
        public bool CheckConditions(ISkillUnlockCharacterState character)
        {
            // Condition checking based on trigger type
            // Full implementation requires the condition evaluator from Phase 2
            return true;
        }

        /// <summary>
        /// Check if character can afford the unlock cost.
        /// Returns (canAfford, reason).
        /// </summary>
        public (bool canAfford, string reason) CheckCost(ISkillUnlockCharacterState character)
        {
            if (!Cost.HasCost)
                return (true, "No cost");

            if (Cost.Gold > 0 && character.Gold < Cost.Gold)
                return (false, $"Need {Cost.Gold} gold (have {character.Gold})");

            if (Cost.SkillPoints > 0 && character.SkillPoints < Cost.SkillPoints)
                return (false, $"Need {Cost.SkillPoints} skill points (have {character.SkillPoints})");

            foreach (var mat in Cost.Materials)
            {
                int count = character.GetItemCount(mat.ItemId);
                if (count < mat.Quantity)
                    return (false, $"Need {mat.Quantity}x {mat.ItemId} (have {count})");
            }

            return (true, "Can afford");
        }
    }

    /// <summary>
    /// Interface for character state needed by the skill unlock system.
    /// </summary>
    public interface ISkillUnlockCharacterState
    {
        int Level { get; }
        int Gold { get; }
        int SkillPoints { get; }
        int GetItemCount(string itemId);
        int GetActivityCount(string activityType);
        List<string> EarnedTitleIds { get; }
        List<string> CompletedQuestIds { get; }
    }

    // ========================================================================
    // Skill Unlock System
    // ========================================================================

    /// <summary>
    /// Manages skill unlock progression for a character.
    /// Tracks which skills have been unlocked and which are pending cost payment.
    /// Matches Python SkillUnlockSystem class exactly.
    /// </summary>
    public class SkillUnlockSystem
    {
        /// <summary>Set of skill IDs that have been unlocked.</summary>
        public HashSet<string> UnlockedSkills { get; } = new();

        /// <summary>Set of unlock IDs awaiting cost payment.</summary>
        public HashSet<string> PendingUnlocks { get; } = new();

        /// <summary>
        /// All available unlock definitions. Set after database is initialized.
        /// </summary>
        public List<SkillUnlockDefinition> AllUnlocks { get; set; } = new();

        // ====================================================================
        // Query Methods
        // ====================================================================

        /// <summary>
        /// Check if a skill has been unlocked.
        /// Matches Python is_skill_unlocked().
        /// </summary>
        public bool IsUnlocked(string skillId)
        {
            return UnlockedSkills.Contains(skillId);
        }

        /// <summary>
        /// Check if an unlock is pending cost payment.
        /// </summary>
        public bool IsPending(string unlockId)
        {
            return PendingUnlocks.Contains(unlockId);
        }

        /// <summary>Get number of skills unlocked.</summary>
        public int UnlockedCount => UnlockedSkills.Count;

        /// <summary>Get list of pending unlock IDs.</summary>
        public List<string> GetPendingUnlocks() => PendingUnlocks.ToList();

        // ====================================================================
        // Unlock Checking
        // ====================================================================

        /// <summary>
        /// Check for new skill unlocks based on character state.
        /// Optionally filter by trigger type and value.
        /// Matches Python check_for_unlocks() exactly.
        /// </summary>
        public List<SkillUnlockDefinition> CheckUnlocks(
            ISkillUnlockCharacterState character,
            string triggerType = null,
            object triggerValue = null)
        {
            var newlyUnlockable = new List<SkillUnlockDefinition>();

            foreach (var unlock in AllUnlocks)
            {
                // Skip already unlocked
                if (UnlockedSkills.Contains(unlock.SkillId))
                    continue;

                // Skip already pending
                if (PendingUnlocks.Contains(unlock.UnlockId))
                    continue;

                // Filter by trigger type if specified
                if (triggerType != null && unlock.Trigger.Type != triggerType)
                    continue;

                // Filter by trigger value if specified
                if (triggerValue != null && !Equals(unlock.Trigger.TriggerValue, triggerValue))
                    continue;

                // Check conditions
                if (!unlock.CheckConditions(character))
                    continue;

                // Conditions met
                var (canAfford, _) = unlock.CheckCost(character);

                if (!unlock.Cost.HasCost)
                {
                    // No cost: unlock immediately
                    UnlockedSkills.Add(unlock.SkillId);
                    newlyUnlockable.Add(unlock);
                }
                else if (canAfford)
                {
                    // Has cost but can afford: mark pending for player confirmation
                    PendingUnlocks.Add(unlock.UnlockId);
                    newlyUnlockable.Add(unlock);
                }
            }

            return newlyUnlockable;
        }

        /// <summary>
        /// Unlock a specific skill by unlock ID (from player UI confirmation).
        /// Matches Python unlock_skill().
        /// </summary>
        public (bool success, string message) UnlockSkill(string unlockId, ISkillUnlockCharacterState character)
        {
            var unlock = AllUnlocks.FirstOrDefault(u => u.UnlockId == unlockId);
            if (unlock == null)
                return (false, $"Unknown unlock: {unlockId}");

            if (UnlockedSkills.Contains(unlock.SkillId))
                return (false, "Skill already unlocked");

            // Check conditions and cost
            if (!unlock.CheckConditions(character))
                return (false, "Conditions not met");

            var (canAfford, reason) = unlock.CheckCost(character);
            if (!canAfford)
                return (false, reason);

            // Unlock (cost payment handled by caller with full character access)
            UnlockedSkills.Add(unlock.SkillId);
            PendingUnlocks.Remove(unlockId);

            return (true, $"Unlocked skill: {unlock.Name}");
        }

        // ====================================================================
        // Trigger-Specific Check Methods
        // ====================================================================

        /// <summary>
        /// Check for unlocks triggered by level up.
        /// Matches Python check_level_up_unlocks().
        /// </summary>
        public List<SkillUnlockDefinition> CheckLevelUpUnlocks(ISkillUnlockCharacterState character, int newLevel)
        {
            return CheckUnlocks(character, triggerType: "level_up", triggerValue: newLevel);
        }

        /// <summary>
        /// Check for unlocks triggered by earning a title.
        /// Matches Python check_title_earned_unlocks().
        /// </summary>
        public List<SkillUnlockDefinition> CheckTitleEarnedUnlocks(ISkillUnlockCharacterState character, string titleId)
        {
            return CheckUnlocks(character, triggerType: "title_earned", triggerValue: titleId);
        }

        /// <summary>
        /// Check for unlocks triggered by quest completion.
        /// Matches Python check_quest_complete_unlocks().
        /// </summary>
        public List<SkillUnlockDefinition> CheckQuestCompleteUnlocks(ISkillUnlockCharacterState character, string questId)
        {
            return CheckUnlocks(character, triggerType: "quest_complete", triggerValue: questId);
        }

        /// <summary>
        /// Check for unlocks triggered by activity thresholds.
        /// Called periodically after crafting, combat, gathering.
        /// Matches Python check_activity_threshold_unlocks().
        /// </summary>
        public List<SkillUnlockDefinition> CheckActivityThresholdUnlocks(ISkillUnlockCharacterState character)
        {
            return CheckUnlocks(character, triggerType: "activity_threshold");
        }

        // ====================================================================
        // Save/Load
        // ====================================================================

        /// <summary>
        /// Serialize to save data.
        /// </summary>
        public Dictionary<string, object> ToSaveData()
        {
            return new Dictionary<string, object>
            {
                ["unlocked_skills"] = UnlockedSkills.ToList(),
                ["pending_unlocks"] = PendingUnlocks.ToList(),
            };
        }

        /// <summary>
        /// Restore from save data.
        /// </summary>
        public void FromSaveData(Dictionary<string, object> data)
        {
            UnlockedSkills.Clear();
            PendingUnlocks.Clear();

            if (data.TryGetValue("unlocked_skills", out var usObj) && usObj is IEnumerable<object> usList)
            {
                foreach (var item in usList)
                    UnlockedSkills.Add(item?.ToString());
            }

            if (data.TryGetValue("pending_unlocks", out var puObj) && puObj is IEnumerable<object> puList)
            {
                foreach (var item in puList)
                    PendingUnlocks.Add(item?.ToString());
            }
        }

        /// <summary>
        /// Reset all unlock state (for debug/testing).
        /// Matches Python reset_debug().
        /// </summary>
        public void ResetDebug()
        {
            UnlockedSkills.Clear();
            PendingUnlocks.Clear();
        }
    }
}
